ufl/core/external_operator.py

```diff
@@ -0,0 +1,278 @@
+        else:
+            corresponding_count = self.count()
+            corresponding_coefficient = coefficient or self._coefficient
+
+        reconstruct_op = type(self)(*operands, function_space=function_space or self.ufl_function_space(),
+                                    derivatives=deriv_multiindex,
+                                    count=corresponding_count,
+                                    coefficient=corresponding_coefficient,
+                                    arguments=arguments or (self.arguments() + self.action_args()))
+
+        if deriv_multiindex != self.derivatives:
+            # If we are constructing a derivative
+            self._extop_master._extop_dependencies.append(reconstruct_op)
+            reconstruct_op._extop_master = self._extop_master
+        else:
+            reconstruct_op._extop_master = self._extop_master
+            reconstruct_op._extop_dependencies = [e._ufl_expr_reconstruct_(*operands) for e in self._extop_dependencies[1:]]
+        return reconstruct_op
+
+    def _add_dependencies(self, derivatives, args):
+        v = list(self._ufl_expr_reconstruct_(*self.ufl_operands, derivatives=d, arguments=a)
+                 for d, a in zip(derivatives, args))
+        self._extop_master._extop_dependencies.extend(x for x in v if x not in self._extop_master._extop_dependencies)
+        return self
+
+    def __repr__(self):
+        "Default repr string construction for operators."
+        # This should work for most cases
+        r = "ExternalOperator(%s; %s, %s)" % (", ".join(repr(op) for op in self.ufl_operands), self._arguments, self._count)
+        return r
+
+    def __str__(self):
+        "Default repr string construction for ExternalOperator operators."
+        # This should work for most cases
+        r = "%s(%s,%s,%s,%s,%s)" % (self._ufl_class_.__name__, ", ".join(repr(op) for op in self.ufl_operands),
+                                    repr(self.ufl_function_space()), repr(self.derivatives), repr(self.ufl_shape),
+                                    repr(self._count))
+        return r
+
+    def _ufl_compute_hash_(self):
+        "Default hash of terminals just hash the repr string."
+        return hash(repr(self))
+
+    def _ufl_signature_data_(self, renumbering):
+        "Signature data for form arguments depend on the global numbering of the form arguments and domains."
+        # Type(self) encompasses the information about the external operator type (GLOBAL or LOCAL)
+        coefficient_signature = self.coefficient._ufl_signature_data_(renumbering)
+        return ("ExternalOperator", self._external_operator_type, *coefficient_signature, *self.derivatives)
+
+    def __eq__(self, other):
+        if not isinstance(other, ExternalOperator):
+            return False
+        if self is other:
+            return True
+        return (self._count == other._count and
+                self._ufl_function_space == other._ufl_function_space)
```

Comment:
Is this basically the same equality that `Coefficient` provides?

Reply:
Yes it is, do you think I should use the coefficient produced by the external operator for this? Something like:

```
def __eq__(self, other):
	if not isinstance(other, ExternalOperator):
	            return False
	return self.coefficient.__eq__(other.coefficient)
```